### PR TITLE
[JIT] check stack size after calling operator

### LINKED ISSUE #66650: [JIT] Add Debug Assertions that the Stack is empty after inference
## 🚀 Feature

It's easy to accidentally leave extra values on the stack. This should be easy to catch - at the end of the invocation the size of the stack must be equal to number of returns. 

Change is in https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/runtime/interpreter.cpp, TODO: figure out more detailed information

### PR DESCRIPTION
**Pull request by davidberard98**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #68788

In debug mode, this should throw errors for ops where the wrong number ops is returned (i.e. the number of values left on the stack is different from the number shown in the schema)

Test Plan:
Run this in debug mode and verify that it doesn't throw an assert
```
import torch

class Thing(torch.nn.Module):
    torch.jit.export
    def en(self, x: torch.Tensor):
        return torch.add(x, 2.0)

    def forward(self, x: torch.Tensor, y: torch.Tensor):
        a = torch.mm(x, y)
        b = torch.nn.functional.gelu(a)
        c = self.en(b)
        return c.std_mean()

if __name__ == '__main__':
    unsc = Thing()
    thing = torch.jit.script(unsc)
    x = torch.randn(4, 4)
    y = torch.randn(4, 4)
    std, mean = thing.forward(x, y)
    print(std, mean)
    print(str(thing.forward.graph))
```

Differential Revision: [D32625256](https://our.internmc.facebook.com/intern/diff/D32625256)

### CODE REVIEW

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32625256).

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32625256).

**Comment by eellison** [comment]
Can you verify that all tests pass with this on in debug ?

**Comment by davidberard98** [comment]
@eellison `build/bin/test_jit` and `python3 tests/test_jit.py` both pass, anything else I should check?

I think one of the CI tests is a debug build (I had some failures when I was working on the pr, but I can't find which test it was)

File: torch/csrc/jit/runtime/interpreter/code_impl.h [hunk 1]
```diff
@@ -759,8 +766,32 @@ struct CodeImpl {
     }
 
     operator_table_.emplace_back(oper);
+#ifndef NDEBUG
+    full_operator_table_.emplace_back(op);
+#endif
     return size;
   }
+
+  void assert_stack_size(
```
Comment (eellison) [inline torch/csrc/jit/runtime/interpreter/code_impl.h:775]: Maybe add inline here, if this didn't get inlined then we would be adding overhead in the non-debug. also `inline` doesn't actually guarantee something gets inlined so maybe its better to be safe and add the macros

**Review by eellison — COMMENTED** [review]
Testing sounds good, one more comment abt making sure we're not adding over head in the non-debug case

File: torch/csrc/jit/runtime/interpreter.cpp [hunk 2]
```diff
@@ -284,13 +284,17 @@ struct InterpreterStateImpl : c10::intrusive_ptr_target {
           }
           case INST(OP): {
             INST_GUARD;
+            size_t init_size = stack.size();
             frame.function->operator_table_[inst.X](stack);
+            frame.function->assert_stack_size(inst.X, init_size, stack.size());
```
[the remark this answers is not in the document]
Comment (davidberard98) [reply torch/csrc/jit/runtime/interpreter.cpp:292]: Yeah, I'll just check the #NDEBUG macro

**Review by eellison — APPROVED** [review]

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32625256).

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32625256).

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32625256).

**Pull request closed without merging**